File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution(object):
    def canReach(self, arr, start):
        return self.buscaProf(arr, start, 0)
    

    # Solução recursiva
    def buscaProf(self, arr, start, contador):
        # Caso start menor que 0 ou satart maior que o tamanho do array
        # Retorna False
        if start < 0 or start >= len(arr):
            return False
        
        # Caso o contador (inicialmente == 0) maior ou igual ao tamanho do array
        # Retorna False
        if contador >= len(arr):
            return False
    
        # Caso o arr[start] == 0
        # Retorna True
        if arr[start] == 0:
            return True

        # Retorna buscando pela direita ou pela esquerda caso nenhum dos casos acima seja verdadeiro
        return self.buscaProf(arr, start + arr[start], contador + 1) or \
               self.buscaProf(arr, start - arr[start], contador + 1)
    
    
    

File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        queue <int> q;
        q.push(start);

        while (!q.empty()){
            int atual = q.front();
            q.pop();

            if (arr[atual] == 0 ){
                return true;
            }

            // verifica se a posição atual já foi visitada anteriormente
            if (arr[atual] < 0){
                // caso já tenha sido verificada anteriormente
                // continua toda a operação
                continue;
            }

            // verifica se está dentro dos limites do array
            // verifica à direita
            if(atual + arr[atual] < arr.size()){
                //adiciona a fila 'q' a nova posição a analisar
                q.push(atual + arr[atual]);
            }

            // verifica se está dentro dos limites do array
            // verifica à esquerda
            if(atual - arr[atual] >= 0){
                //adiciona a fila 'q' a nova posição a analisar
                q.push(atual - arr[atual]);
            }

            // marca que aquela posição já foi visitada 
            arr[atual] = -arr[atual];
        }

        return false;

    }
};